app/lib/shared/exceptions.dart

```diff
@@ -2,21 +2,223 @@
+abstract class ResponseException extends ApiResponseException {
+  ResponseException._(int status, String code, String message)
+      : super(status: status, code: code, message: message);
+
+  @override
+  String toString() => '$code($status): $message'; // implemented for debugging
+}
+
+/// Thrown when resource does not exist.
+class NotFoundException extends ResponseException {
+  NotFoundException(String message) : super._(404, 'NotFound', message);
+  NotFoundException.resource(String resource)
+      : super._(404, 'NotFound', 'Could not find `$resource`.');
+}
+
+/// Thrown when request input is invalid, bad payload, wrong querystring, etc.
+class InvalidInputException extends ResponseException {
+  InvalidInputException._(String message)
+      : super._(
+          400,
+          'InvalidInput', // also duplicated in api_builder.dart
+          message,
+        );
+
+  /// Check [condition] and throw [InvalidInputException] with [message] if
+  /// [condition] is `false`.
+  static void check(bool condition, String message) {
+    assert(message != null, '"message" must not be `null`');
+    if (!condition) {
+      throw InvalidInputException._(message);
+    }
+  }
+
+  /// A variant of [check] with lazy message construction.
+  static void _check(bool condition, String Function() message) {
+    assert(message != null, '"message" creator must not be `null`');
+    assert(message() != null, '"message()" creator must not return `null`');
+    if (!condition) {
+      throw InvalidInputException._(message());
+    }
+  }
+
+  /// Throw [InvalidInputException] if [value] is `null`.
+  static void checkNotNull(dynamic value, String name) {
+    assert(name != null, '"name" must not be `null`');
+    _check(value != null, () => '"$name" cannot be `null`');
+  }
+
+  /// Throw [InvalidInputException] if [value] doesn't match [pattern].
+  static void checkMatchPattern(String value, String name, Pattern pattern) {
+    assert(name != null, '"name" must not be `null`');
+    assert(pattern != null, '"pattern" must not be `null`');
+    _check(pattern.allMatches(value).isNotEmpty,
+        () => '"$name" must match $pattern');
+  }
+
+  /// Throw [InvalidInputException] if [value] is not one of [values].
+  static void checkAnyOf<T>(T value, String name, Iterable<T> values) {
+    assert(name != null, '"name" must not be `null`');
+    assert(values != null, '"values" must not be `null`');
+    _check(values.contains(value),
+        () => '"$name" must be any of ${values.join(', ')}');
+  }
+
+  /// Throw [InvalidInputException] if [value] is less than [minimum] or greater
+  /// than [maximum].
+  static void checkRange<T extends num>(
+    T value,
+    String name, {
+    T minimum,
+    T maximum,
+  }) {
+    _check(value != null, () => '"$name" cannot be `null`');
+    assert(name != null, '"name" must not be `null`');
+    _check((minimum == null || value >= minimum),
```

Comment:
It is a bit confusing to read this, I was excepting to throw the exception when the condition is true. Maybe `_assertOrThrow`?

Reply:
Checks that the condition is true... maybe we could name it `_ensure` but it's really named after the public variant..